File: sdks/kotlin/src/test/kotlin/com/sdk_tpl/CircuitBreakerTest.kt
package com.sdk_tpl

import com.sdk_tpl.config.CircuitBreakerConfig
import com.sdk_tpl.errors.ErrorCode
import com.sdk_tpl.errors.SdkTplException
import com.sdk_tpl.http.CircuitBreaker
import kotlinx.coroutines.test.runTest
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith

class CircuitBreakerTest {

    @Test
    fun `circuit breaker starts in closed state`() = runTest {
        val cb = CircuitBreaker(CircuitBreakerConfig())
        assertEquals(CircuitBreaker.State.CLOSED, cb.currentState())
    }

    @Test
    fun `successful calls keep circuit closed`() = runTest {
        val cb = CircuitBreaker(CircuitBreakerConfig(failureThreshold = 3))

        val result = cb.execute { "success" }

        assertEquals("success", result)
        assertEquals(CircuitBreaker.State.CLOSED, cb.currentState())
        assertEquals(0, cb.currentFailureCount())
    }

    @Test
    fun `failures below threshold keep circuit closed`() = runTest {
        val cb = CircuitBreaker(CircuitBreakerConfig(failureThreshold = 3))

        repeat(2) {
            try {
                cb.execute {
                    throw SdkTplException.networkError("fail")
                }
            } catch (_: SdkTplException) { }
        }

        assertEquals(CircuitBreaker.State.CLOSED, cb.currentState())
        assertEquals(2, cb.currentFailureCount())
    }

    @Test
    fun `circuit opens after reaching failure threshold`() = runTest {
        val cb = CircuitBreaker(CircuitBreakerConfig(failureThreshold = 3))

        repeat(3) {
            try {
                cb.execute {
                    throw SdkTplException.networkError("fail")
                }
            } catch (_: SdkTplException) { }
        }

        assertEquals(CircuitBreaker.State.OPEN, cb.currentState())
    }

    @Test
    fun `open circuit rejects requests`() = runTest {
        val cb = CircuitBreaker(CircuitBreakerConfig(failureThreshold = 1))

        try {
            cb.execute { throw SdkTplException.networkError("fail") }
        } catch (_: SdkTplException) { }

        val exception = assertFailsWith<SdkTplException> {
            cb.execute { "should not execute" }
        }

        assertEquals(ErrorCode.CIRCUIT_BREAKER_OPEN, exception.errorCode)
    }

    @Test
    fun `non-recoverable errors do not count as failures`() = runTest {
        val cb = CircuitBreaker(CircuitBreakerConfig(failureThreshold = 2))

        repeat(5) {
            try {
                cb.execute {
                    throw SdkTplException.authenticationError()
                }
            } catch (_: SdkTplException) { }
        }

        assertEquals(CircuitBreaker.State.CLOSED, cb.currentState())
        assertEquals(0, cb.currentFailureCount())
    }

    @Test
    fun `successful call after failures resets count`() = runTest {
        val cb = CircuitBreaker(CircuitBreakerConfig(failureThreshold = 5))

        repeat(3) {
            try {
                cb.execute { throw SdkTplException.networkError("fail") }
            } catch (_: SdkTplException) { }
        }

        assertEquals(3, cb.currentFailureCount())

        cb.execute { "success" }

        assertEquals(0, cb.currentFailureCount())
        assertEquals(CircuitBreaker.State.CLOSED, cb.currentState())
    }

    @Test
    fun `reset restores circuit to initial state`() = runTest {
        val cb = CircuitBreaker(CircuitBreakerConfig(failureThreshold = 1))

        try {
            cb.execute { throw SdkTplException.networkError("fail") }
        } catch (_: SdkTplException) { }

        assertEquals(CircuitBreaker.State.OPEN, cb.currentState())

        cb.reset()

        assertEquals(CircuitBreaker.State.CLOSED, cb.currentState())
        assertEquals(0, cb.currentFailureCount())
    }
}

File: sdks/php/tests/CircuitBreakerTest.php
<?php

declare(strict_types=1);

namespace SdkTpl\Tests;

use PHPUnit\Framework\TestCase;
use SdkTpl\Errors\ErrorCode;
use SdkTpl\Errors\SdkTplException;
use SdkTpl\Http\CircuitBreaker;

class CircuitBreakerTest extends TestCase
{
    public function testStartsInClosedState(): void
    {
        $cb = new CircuitBreaker();
        $this->assertSame(CircuitBreaker::STATE_CLOSED, $cb->getState());
    }

    public function testSuccessfulCallsKeepCircuitClosed(): void
    {
        $cb = new CircuitBreaker(failureThreshold: 3);

        $result = $cb->execute(fn () => 'success');

        $this->assertSame('success', $result);
        $this->assertSame(CircuitBreaker::STATE_CLOSED, $cb->getState());
        $this->assertSame(0, $cb->getFailureCount());
    }

    public function testFailuresBelowThresholdKeepCircuitClosed(): void
    {
        $cb = new CircuitBreaker(failureThreshold: 3);

        for ($i = 0; $i < 2; $i++) {
            try {
                $cb->execute(function (): never {
                    throw SdkTplException::networkError('fail');
                });
            } catch (SdkTplException) {
            }
        }

        $this->assertSame(CircuitBreaker::STATE_CLOSED, $cb->getState());
        $this->assertSame(2, $cb->getFailureCount());
    }

    public function testCircuitOpensAfterReachingFailureThreshold(): void
    {
        $cb = new CircuitBreaker(failureThreshold: 3);

        for ($i = 0; $i < 3; $i++) {
            try {
                $cb->execute(function (): never {
                    throw SdkTplException::networkError('fail');
                });
            } catch (SdkTplException) {
            }
        }

        $this->assertSame(CircuitBreaker::STATE_OPEN, $cb->getState());
    }

    public function testOpenCircuitRejectsRequests(): void
    {
        $cb = new CircuitBreaker(failureThreshold: 1);

        try {
            $cb->execute(function (): never {
                throw SdkTplException::networkError('fail');
            });
        } catch (SdkTplException) {
        }

        try {
            $cb->execute(fn () => 'should not execute');
            $this->fail('Expected SdkTplException to be thrown');
        } catch (SdkTplException $e) {
            $this->assertSame(ErrorCode::CIRCUIT_BREAKER_OPEN, $e->getErrorCode());
        }
    }

    public function testNonRecoverableErrorsDoNotCountAsFailures(): void
    {
        $cb = new CircuitBreaker(failureThreshold: 2);

        for ($i = 0; $i < 5; $i++) {
            try {
                $cb->execute(function (): never {
                    throw SdkTplException::authenticationError();
                });
            } catch (SdkTplException) {
            }
        }

        $this->assertSame(CircuitBreaker::STATE_CLOSED, $cb->getState());
        $this->assertSame(0, $cb->getFailureCount());
    }

    public function testSuccessfulCallAfterFailuresResetsCount(): void
    {
        $cb = new CircuitBreaker(failureThreshold: 5);

        for ($i = 0; $i < 3; $i++) {
            try {
                $cb->execute(function (): never {
                    throw SdkTplException::networkError('fail');
                });
            } catch (SdkTplException) {
            }
        }

        $this->assertSame(3, $cb->getFailureCount());

        $cb->execute(fn () => 'success');

        $this->assertSame(0, $cb->getFailureCount());
        $this->assertSame(CircuitBreaker::STATE_CLOSED, $cb->getState());
    }

    public function testResetRestoresCircuitToInitialState(): void
    {
        $cb = new CircuitBreaker(failureThreshold: 1);

        try {
            $cb->execute(function (): never {
                throw SdkTplException::networkError('fail');
            });
        } catch (SdkTplException) {
        }

        $this->assertSame(CircuitBreaker::STATE_OPEN, $cb->getState());

        $cb->reset();

        $this->assertSame(CircuitBreaker::STATE_CLOSED, $cb->getState());
        $this->assertSame(0, $cb->getFailureCount());
    }
}
